### PR TITLE
Take advantage of Gradle 8 cache cleanup

### PR DESCRIPTION
**Pull request by skjolber**
Key feature of incremental cache: prevent the cache from growing too big.

Gradle 8 can clean up dependencies after n days. Deprecate the custom gradle code made for version 7 and below.

Implementation for Gradle 8: 

 * disable GC run while steps are executing
 * force GC before cache is persisted 

This still keeps 1 day old items around, forever, untill a new build change is made. To sum up:

 * cache is restored on build file changes
 * cache restore is marginally slower on "non build file changes"
 
 Future improvement: Flush cache on first commit after 1 day.

### CODE REVIEW

**Comment by skjolber** [comment]
Optimization: Let gradle clean up the cache after 2 days, store the result under seperate cache keys.

File: .circleci/test-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  orb-tools: circleci/orb-tools@11.6.1
+filters: &filters
+  tags:
+    only: /.*/
+jobs:
+  # This job is an example of an integration testing job.
+  # This job should execute a command from your orb and verify
+  # the output is as expected, otherwise the job should fail.
+  #
+  # Rename this job and add more as needed.
```
Comment (tommybo-entur) [inline .circleci/test-deploy.yml:13]: These comments should probably be updated or removed.

File: .circleci/test-deploy.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  #
+  integration-test-1:
```
Comment (tommybo-entur) [inline .circleci/test-deploy.yml:15]: This test doesn't do much, it's more of an expectation test that the build image has gradle installed. It's debatable if this is necessary for the `gradle/default` executor.

File: .circleci/test-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    executor: gradle/default
+    steps:
+      - checkout
+      - run:
+          name: Check if Gradle is installed
```
Comment (tommybo-entur) [inline .circleci/test-deploy.yml:20]: Perhaps a couple of tests to verify the cashing behavior would be in order?

File: README.md [hunk 4]
```diff
@@ -14,12 +10,11 @@ Typical use-case:
 Advantages:
 
  * Improves build time when the Gradle build files themselves are updated, as (most) dependencies are already in the cache
- * Less traffic for artifact repositories (i.e. Maven central, JCenter, your own private Artifactory etc.)
+ * Much less traffic for artifact repositories (i.e. Maven central, JCenter, your own private Artifactory etc.)
```
Comment (tommybo-entur) [inline README.md:13]: Remove JCenter from the examples, perhaps?

File: README.md [hunk 5]
```diff
@@ -14,12 +10,11 @@ Typical use-case:
...
 
 Disadvantages:
 
- *  Uses internal Gradle classes to delete unused dependencies. 
-      * The cache job might break on a future version of the Gradle wrapper. If so, it is trivial to (temporarily) revert to the official Gradle orb.
+ *  Uses internal Gradle classes to delete unused dependencies for Gradle versions < 8. For Gradle version >= 8 it uses the built in cache cleanup strategy (with 1 day retention of unused dependencies).
```
Comment (tommybo-entur) [inline README.md:17]: Is it better to use an earlier version of this orb for Gradle < 8?

File: src/examples/custom_reports.yml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+usage:
+  version: 2.1
+
+  orbs:
+    gradle: entur/gradle-orb@x.y
+
+  workflows:
+    test-with-custom-reports:
+      jobs:
+        - gradle/test:
```
Comment (tommybo-entur) [inline src/examples/custom_reports.yml:12]: Perhaps we should use values that differs from default in the example?

File: src/scripts/checksum_build_files.sh
Comment (tommybo-entur) [inline src/scripts/checksum_build_files.sh:1]: This looks like it could use an unit test.

File: src/scripts/cleanup.sh
Comment (tommybo-entur) [inline src/scripts/cleanup.sh:1]: This also looks like it could use an unit test.

File: src/scripts/cleanup.sh [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    ./gradlew --stop
+    ./gradlew -b /tmp/cleanup.gradle dummy --no-daemon --info
+    # for debugging
+    du -h --max-depth=1 "$GRADLE_CACHE_DIRECTORY"
+    exit 0
+fi
+# this is the first successful build with this particular set of build files
+echo "Clean up cache for gradle < 8"
+touch /tmp/settings.gradle
+cat > /tmp/cleanup.gradle << 'endmsg'
```
Comment (tommybo-entur) [inline src/scripts/cleanup.sh:59]: perhaps cleanup.gradle should be added as a separate file?

**Review by tommybo-entur — COMMENTED** [review]
All in all: Great work. Left a few comments to consider.

File: .circleci/config.yml [hunk 8]
```diff
@@ -1,40 +1,37 @@
...
     jobs:
-      - lint
+      - orb-tools/lint:
+          filters: *filters
+      - orb-tools/pack:
+          filters: *filters
+      - orb-tools/review:
+          filters: *filters
+      - shellcheck/check:
+          exclude: SC2148,SC2038,SC2086,SC2002,SC2016
```
[the remark this answers is not in the document]
Comment (skjolber) [reply .circleci/config.yml:21]: Good question, this code just copied from the circleci gradle orb

File: src/scripts/cleanup.sh (hunk 7 above)
Comment (skjolber) [reply src/scripts/cleanup.sh:59]: Yes, but it is not possible to include files in the orb (technically all imports are embedded).

File: README.md [hunk 9]
```diff
@@ -1,113 +1 @@
...
-
-
-## Troubleshooting
-If the cache is corrupted, update the cache key, so that the previous state is not restored - as in the official Gradle orb.
-
-[issue-tracker]:               https://github.com/entur/gradle-orb
-[CircleCI Gradle]:             https://circleci.com/orbs/registry/orb/circleci/gradle?version=2.2.0
-
-
+Test project
```
Comment (tommybo-entur) [inline README.md:1]: Wrong file? Was this change intended for `sample_app/README.md`?

Comment (skjolber) [reply README.md:1]: Reverting

File: README.md (hunk 5 above)
Comment (skjolber) [reply README.md:17]: No, not really. I did consider just supporting gradle 8+ in the latest version, but supporting all versions in the same orb makes for a smoother transistion.

File: src/commands/with_cache.yml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  present in the `working_directory`.
+parameters:
+  steps:
+    type: steps
+  app_src_directory:
+    description: Useful when the source of your gradle project is not in the root directory of your git repo. Supply the name of the directory or relative path of the directory containing your source code.
+    type: string
+    default: ''
+  cache_key:
+    description: Add a custom suffix to your cache key in the event you need to work with multiple maven caches.
```
[the remark this answers is not in the document]
Comment (skjolber) [reply src/commands/with_cache.yml:18]: Yes

File: src/scripts/cleanup.sh
Comment (tommybo-entur) [reply src/scripts/cleanup.sh:1]: Just unit test the shell script. No need to run it as an integration test.

File: src/examples/custom_reports.yml (hunk 6 above)
Comment (skjolber) [reply src/examples/custom_reports.yml:12]: Will revisit this in next version, use of these parameters + `app_src_directory` is not 100%

**Review by Chr1stian — APPROVED** [review]

**Review by tommybo-entur — APPROVED** [review]
LGTM

File: src/scripts/cleanup.sh
Comment (skjolber) [reply src/scripts/cleanup.sh:1]: Test added
